site/src/backend/auth.js

```diff
@@ -0,0 +1,149 @@
+import polka from 'polka';
+import devalue from 'devalue';
+import send from '@polka/send';
+import { get, post } from 'httpie';
+import { parse, stringify } from 'querystring';
+import { decode, sign, verify } from './token';
+import { find, query } from '../utils/db';
+
+const {
+	BASEURL,
+	GITHUB_CLIENT_ID,
+	GITHUB_CLIENT_SECRET,
+} = process.env;
+
+const OAuth = 'https://github.com/login/oauth';
+
+function exit(res, code, msg='') {
+	const error = msg.charAt(0).toUpperCase() + msg.substring(1);
+	send(res, code, { error });
+}
+
+function onError(err, req, res) {
+	const error = err.message || err;
+	const code = err.code || err.status || 500;
+	res.headersSent || send(res, code, { error });
+}
+
+/**
+ * Middleware to determine User validity
+ */
+export async function isUser(req, res) {
+	const abort = exit.bind(null, res, 401);
+
+	const auth = req.headers.authorization;
+	if (!auth) return abort('Missing Authorization header');
+
+	const [scheme, token] = auth.split(' ');
+	if (scheme !== 'Bearer' || !token) return abort('Invalid Authorization format');
+
+	let data;
+	const decoded = decode(token, { complete:true });
+	if (!decoded || !decoded.header) return abort('Invalid token');
+
+	try {
+		data = await verify(token);
+	} catch (err) {
+		return abort(err.message);
+	}
+
+	const { uid, username } = data;
+	if (!uid || !username) return abort('Invalid token payload');
+
+	try {
+		const row = await find(`select * from users where uid = $1 and username = $2 limit 1`, [uid, username]);
+		return row || abort('Invalid token');
+	} catch (err) {
+		console.error('Auth.isUser', err);
+		return send(res, 500, 'Unknown error occurred');
+	}
+}
+
+export function toUser(obj={}) {
+	const { uid, username, name, avatar } = obj;
+	const token = sign({ uid, username });
+	return { uid, username, name, avatar, token };
```

Comment:
This is the only user data that ever makes it to the client. The `uid` is GitHub's internal ID for the user. Not sure how "okay" it to keep this in localStorage.

Currently we don't. See comment above.